### ISSUE TITLE
Implement line-height quirks

### ISSUE THREAD

**Issue by sole** [servo/servo#11704]
My test at http://sole.github.io/test_cases/servo/gaps/ renders a seamless 'checkerboard' in Firefox, Safari and Chrome:
<img width="1064" alt="screen shot 2016-06-10 at 13 13 23" src="https://cloud.githubusercontent.com/assets/5609/15963913/6894ee0e-2f0d-11e6-86da-500596925589.png">

but it renders an unexpected gap under each row in Servo:
<img width="912" alt="screen shot 2016-06-10 at 13 08 53" src="https://cloud.githubusercontent.com/assets/5609/15963920/75d5ead2-2f0d-11e6-908b-b322acd2fa11.png">

I have a hunch this might be because Servo is assuming doctype=html and my test case _doesn't_ include the doctype. When adding the doctype, all the browsers render the gap (which is unexpected anyway, but hey, CSS `¯\_(ツ)_/¯`)

**Comment by Ms2ger** [comment]
```html
<html>
	<head>
		<meta charset="utf-8">
		<title>strange gaps</title>
	</head>
	<body>
		<p>there should be no gaps between each row of checkers</p>
		<div id="checkers"></div>
		<style>
			#checkers > div {
				background: #000;
				display: inline-block;
				width: 10vw;
				height: 10vw;
				transition: all 0.5s ease-in-out;
			}
			
			#checkers div:nth-child(odd) {
				background: #f00;
			}
		</style>

		<script>
			window.onload = function() {

				var checkers = document.getElementById('checkers');
				for(var i = 0; i < 36; i++) {
					var el = document.createElement('div');
					checkers.appendChild(el);
				}

			};
		</script>
	</body>
</html>
```

This is either https://quirks.spec.whatwg.org/#the-line-height-calculation-quirk or https://quirks.spec.whatwg.org/#the-blocks-ignore-line-height-quirk

**Comment by SimonSapin** [comment]
## What’s going on

https://drafts.csswg.org/css2/visudet.html#propdef-line-height
> On a block container element whose content is composed of inline-level elements, 'line-height' specifies the minimal height of line boxes within the element. The minimum height consists of a minimum height above the baseline and a minimum depth below it, exactly as if each line box starts with a zero-width inline box with the element's font and line height properties. We call that imaginary box a "strut." 

`<div id="checkers">` is such a block container. Let’s assume that `font-size` is 16px, `line-height` is 1.2, and the first available font in `font-family` places the baseline 0.2em from the bottom of its EM box. The strut will be 19.2px high, of which 15.36px is above the baseline and 3.84px below.

Additionally, the initial `vertical-align` is `baseline`, so the baseline of the strut is aligned with that of the inner divs:

> The baseline of an 'inline-block' is the baseline of its last line box in the normal flow, unless it has either no in-flow line boxes or if its 'overflow' property has a computed value other than 'visible', in which case the baseline is the bottom margin edge.

The inner divs are empty, so their baseline is their bottom edge. They go 10vw (let’s say that’s 100px) above the baseline and 0px below.

Overall, the height of the line is that of the bounding box of everything in it. In this case it’s `max(15.36px, 100px) + max(3.84px, 0px)` which is 103.84px. (Some of these values may be clamped to device pixels in implementations.)

## How to work around it for authors

When `line-height` is set to 0, the strut becomes zero-height and so only the inner divs take up vertical space in the line.

Another way to remove the gap, if the inner divs are higher than the strut, is to use something like `vertical-align: top` so that the baseline is not relevant.

## What to do in Servo

As documented in the spec linked above by Ms2ger, in quirks mode other browsers set `line-height` to 0 in some cases. Let’s make this issue about implementing this.

CC https://github.com/servo/servo/issues/11709

**Comment by notriddle** [comment]
This is the same bug as #10726, only for inline-block elements instead of canvases (it also applies to images).

**Comment by notriddle** [comment]
The checkerboard test in the test comment should match this:

```html
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>strange gaps reference</title>
    </head>
    <body>
        <p>there should be no gaps between each row of checkers</p>
        <div id="checkers"></div>
        <style>
            #checkers > div {
                background: #000;
                display: inline-block;
                width: 10vw;
                height: 10vw;
                float: left;
            }

            #checkers div:nth-child(odd) {
                background: #f00;
            }
        </style>

        <script>
            window.onload = function() {

                var checkers = document.getElementById('checkers');
                for(var i = 0; i < 36; i++) {
                    var el = document.createElement('div');
                    checkers.appendChild(el);
                }

            };
        </script>
    </body>
</html>
```

Note that I'm using floats, so servo-master will render it right.

**Comment by jimberlage** [comment]
Mind if I tackle this?

**Comment by jdm** [comment]
This might be what https://github.com/servo/servo/pull/11853 is addressing, but I'm not sure. @nox, do you know?

**Comment by zcorpan** [comment]
There are two quirks here that are related

https://quirks.spec.whatwg.org/#the-line-height-calculation-quirk (#11853)
https://quirks.spec.whatwg.org/#the-blocks-ignore-line-height-quirk (this issue? the gap is caused by the "strut")

In [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Mozilla_quirks_mode_behavior#Block_and_Inline_layout) these two are described as one and the same.

**Comment by nox** [comment]
@jdm Yes that's what I'm *trying* to address, with difficulty.

**Comment by pshaughn** [comment]
In addition to tests that are specifically about this quirk, WPT also uses it as an am-I-in-quirks-mode checker for tests that are asking whether quirks mode is happening, like html/webappapis/dynamic-markup-insertion/opening-the-input-stream/quirks.window

**Comment by dralley** [comment]
This issue looks fixed to me, should it be closed?

![Uploading Screenshot from 2020-04-01 16-27-44.png…]()

**Comment by zcorpan** [comment]
Do the following tests in http://wpt.live/quirks/ pass?
```
blocks-ignore-line-height.html
line-height-calculation.html
line-height-in-list-item.tentative.html
line-height-trailing-collapsable-whitespace.html
```

**Comment by jdm** [comment]
Nope: https://github.com/servo/servo/tree/master/tests/wpt/metadata/quirks

**Comment by lumiscosity** [comment]
The mentioned tests now pass, and the checkers example (both testcases provided here) still works, so the issue can probably be closed. Tested on Arch Linux x86_64, servo built against commit id 9d29017c0dc6eec4ca7cd868cc13056e0bc0c64a, compared against Firefox 141.0, Epiphany 48.5 and Chromium 138.0.7204.168 (ordered left to right).

<img width="1814" height="560" alt="Image" src="https://github.com/user-attachments/assets/090f715a-3b54-4968-af6d-a645682bc5fc" />

**Comment by jdm** [comment]
Thanks for verifying!
